data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateProcessor.java

```diff
@@ -19,23 +19,54 @@
 @DataPrepperPlugin(name = "aggregate", pluginType = Processor.class, pluginConfigurationType = AggregateProcessorConfig.class)
 public class AggregateProcessor extends AbstractProcessor<Record<Event>, Record<Event>> {
     private static final Logger LOG = LoggerFactory.getLogger(AggregateProcessor.class);
 
     private final AggregateProcessorConfig aggregateProcessorConfig;
+    private final GroupStateManager groupStateManager;
+    private final AggregateAction aggregateAction;
+    private final AggregateIdentificationKeysHasher aggregateIdentificationKeysHasher;
 
     @DataPrepperPluginConstructor
-    public AggregateProcessor(final AggregateProcessorConfig aggregateProcessorConfig, final PluginMetrics pluginMetrics, final PluginFactory pluginFactory, final PipelineDescription pipelineDescription) {
+    public AggregateProcessor(final AggregateProcessorConfig aggregateProcessorConfig, final PluginMetrics pluginMetrics, final PluginFactory pluginFactory) {
+        this(aggregateProcessorConfig, pluginMetrics, pluginFactory, new GroupStateManager(), new AggregateIdentificationKeysHasher());
+    }
+    public AggregateProcessor(final AggregateProcessorConfig aggregateProcessorConfig, final PluginMetrics pluginMetrics, final PluginFactory pluginFactory, final GroupStateManager groupStateManager, final AggregateIdentificationKeysHasher aggregateIdentificationKeysHasher) {
         super(pluginMetrics);
         this.aggregateProcessorConfig = aggregateProcessorConfig;
+        this.groupStateManager = groupStateManager;
+        this.aggregateIdentificationKeysHasher = aggregateIdentificationKeysHasher;
+        this.aggregateAction = loadAggregateAction(pluginFactory);
+    }
+
+    private AggregateAction loadAggregateAction(final PluginFactory pluginFactory) {
+        final PluginModel actionConfiguration = aggregateProcessorConfig.getAggregateAction();
+        final PluginSetting actionPluginSetting = new PluginSetting(actionConfiguration.getPluginName(), actionConfiguration.getPluginSettings());
+        return pluginFactory.loadPlugin(AggregateAction.class, actionPluginSetting);
     }
 
     @Override
     public Collection<Record<Event>> doExecute(Collection<Record<Event>> records) {
-        return records;
+        final List<Record<Event>> recordsOut = new LinkedList<>();
+
+        for (final Record<Event> record : records) {
+            final Event event = record.getData();
+            final Map<Object, Object> identificationKeysHash = aggregateIdentificationKeysHasher.createIdentificationKeyHashFromEvent(event, aggregateProcessorConfig.getIdentificationKeys());
+            final Map<Object, Object> groupStateForEvent = groupStateManager.getGroupState(identificationKeysHash);
+
+            final AggregateActionResponse handleEventResponse = aggregateAction.handleEvent(event, groupStateForEvent);
+
+            final Optional<Event> aggregateActionResponseEvent = Optional.ofNullable(handleEventResponse.getEvent());
```

Comment:
I don't think there is much value in wrapping this into an `Optional` just to use `ifPresent`. I think using an if is better here:

```
Event responseEvent = handleEventResponse.getEvent()
if(responseEvent != null)
  recordsOut.add(new Record<>(responseEvent, record.getMetadata()));
```

- [ ] Remove unnecessary Optional